scipy/sparse/csgraph/_disjoint_set.py

```diff
@@ -0,0 +1,138 @@
+"""
+Disjoint set data structure
+"""
+
+
+import numpy as np
+
+
+class DisjointSet:
+    """ Disjoint set data structure for incremental connectivity queries.
+
+    .. versionadded: 1.6.0
+
+    Attributes
+    ----------
+    n_nodes : int
+        The number of nodes in the set.
+    n_components : int
+        The number of components/subsets.
+
+    Methods
+    -------
+    union
+    find
+
+    Notes
+    -----
+    This class implements the disjoint set [1]_, also known as the *union-find*
+    data structure. The *find* method implements the *path compression*
+    variant. The *union* method implements the *union by size* variant.
+
+    References
+    ----------
+    .. [1] https://en.wikipedia.org/wiki/Disjoint-set_data_structure
+
+    See Also
+    --------
+    connected_components : Analyze the connected components of a sparse graph.
+
+    Examples
+    --------
+    >>> from scipy.sparse.csgraph import DisjointSet
+
+    Initialize a disjoint set:
+
+    >>> dis = DisjointSet()
+
+    Merge some subsets:
+
+    >>> dis.union(0, 1)
+    True
+    >>> dis.union(2, 3)
+    True
+    >>> dis.union(3, 3)
+    False
+
+    Find a root node:
+
+    >>> dis.find(1)
+    0
+
+    """
+    def __init__(self):
+        self.n_nodes = 0
+        self.n_components = 0
+        self._sizes = {}
+        self._parents = {}
+
+    def find(self, x):
+        """Find the root node of `x`.
+
+        Parameters
+        ----------
+        x : int
+            Input node.
+
+        Returns
+        -------
+        root : int
+            Root node of `x`.
+        """
+        if not np.issubdtype(type(x), np.integer):
+            raise TypeError("`x` must be an integer")
+
+        if x not in self._parents:
+            self._sizes[x] = 1
+            self._parents[x] = x
+            self.n_nodes += 1
+            self.n_components += 1
+            return x
+
+        parents = self._parents
+        parent = parents[x]
+        while parent != parents[parent]:
+            parent = parents[parent]
+        parents[x] = parent
+        return parent
+
+    def union(self, a, b):
+        """Merge the subsets of `a` and `b`.
+
+        The smaller subset (the child) is merged into the larger subset (the
+        parent). If the subsets are of equal size, the parent is determined by
+        subset root with the smallest index.
+
+        Parameters
+        ----------
+        a, b : int
+            Node indices to merge.
+
+        Returns
+        -------
+        merged : bool
+            `True` if `a` and `b` were in disjoint sets, `False` otherwise.
+        """
+        if not np.issubdtype(type(a), np.integer):
+            raise TypeError("`a` must be an integer")
+        if not np.issubdtype(type(b), np.integer):
+            raise TypeError("`b` must be an integer")
+
+        a = self.find(a)
+        b = self.find(b)
+        if a == b:
+            return False
+
+        if b < a:
```

Comment:
These two lines look redundant: The relative values of `a` and `b` never play a role; only their corresponding relative values in `self._sizes`.

Reply:
But you could do the swap below and shave off a few lines; that is, have the entire thing boil down to:
```python
a = self.find(a)
b = self.find(b)
if a == b: return False

sizes = self._sizes
parents = self._parents
if sizes[a] < sizes[b]:
    a, b = b, a

parents[b] = a
sizes[a] += sizes[b]

self.n_components -= 1
return True
```

Reply:
Ah, I now see that the first swap helps you break ties. (Is that important?)

Reply:
I want a tie-breaker in case `self._sizes[a] == self._sizes[b].
I have documented this in the docstring with the text

> If the subsets are of equal size, the parent is determined by subset root with the smallest index.

Reply:
Sorry, I should have reloaded this page before commenting.

I prefer it if ties are broken. It means that `union(a, b)` always produces the same behaviour as `union(b, a)`. I think I will leave it as-is. What do you think? When I do a C++ V2 this will be almost free.

Reply:
Heh, looks like I just managed to squeeze in my realization the same minute as you wrote the explanation. But the logic sounds reasonable enough to me! It takes some flexibility away from the API, but I don't see a case where that's really an issue.

The phrasing of the last sentence in the documentation does seem ever so slightly off to me; could something like "If the subsets are of equal size, the parent is determined to be the subset whose root has the smallest index" be a bit clearer?

Reply:
I got rid of tie-breaking since I have changed the interface to allow arbitrary objects. Equality is difficult enough, so I don't want to start dealing with inequalities.